base/rational.jl

```diff
@@ -129,12 +129,220 @@ Bool(x::Rational) = x==0 ? false : x==1 ? true :
+    # `excess_precision == max(1, max(num_bit_width, den_bit_width) - 1)`
+    excess_precision = bit_width(quo_) - requested_precision
+
+    # Normalized exponent
+    nexp = bit_width(quo_) - bit_shift_num_ - true
+
+    # Take possible underflow into account: if there's underflow, the
+    # precision needs to be less than requested.
+    adjusted_precision = requested_precision
+    if !isnothing(exp_lower_bound)
+        underflow = clamped_to_zero(exp_lower_bound - nexp)
+        adjusted_precision -= underflow
+    end
+    adjusted_excess_precision = bit_width(quo_) - adjusted_precision
+
+    (adjusted_precision < false) && return FloatComponentsResult(
+        zero(quo_), zero(nexp), inexact = true, underflowed = true)
+
+    # Creates a mantissa in `quo` that's exactly `adjusted_precision`
+    # bits wide, except if rounding up happens, in which case the bit
+    # width may be `adjusted_precision + 1`.
+    bit_shift_num = clamped_to_zero(bit_shift_num_ - adjusted_excess_precision)
+    bit_shift_den = adjusted_excess_precision - (bit_shift_num_ - bit_shift_num)  # nonnegative
+    (quo, rem) = divrem(numT << bit_shift_num, denT << bit_shift_den, romo)
+
+    result_is_exact = iszero(rem)
+
+    nexp_final = nexp + (adjusted_precision < bit_width(quo))
+    is_subnormal = !isnothing(exp_lower_bound) && (nexp_final < exp_lower_bound)
+
+    iszero(quo) || (quo >>= trailing_zeros(quo))
+
+    # The bit width of `quo` is now less than or equal to
+    # `adjusted_precision`, except if `adjusted_precision` is zero, in
+    # which case `bit_width(quo)` may be one, in case rounding up
+    # happened.
+
+    FloatComponentsResult(
+        quo, nexp_final, inexact = !result_is_exact, underflowed = is_subnormal)
+end
+
+# `num`, `den` are positive integers. `bit_width` is the requested
+# floating-point precision and must be positive. `T` is the integer
+# type that we'll be working with mostly, it needs to be wide enough.
+# `exp_lower_bound` is the minimum allowed normalized exponent for
+# normal numbers.
+function rational_to_float_components(num::Integer,
+                                      den::Integer,
+                                      bit_width::Integer,
+                                      ::Type{T},
+                                      romo::RoundingMode,
+                                      exp_lower_bound::Union{Nothing,Integer} =
+                                          nothing) where {T<:Integer}
+    h(::Nothing, ::Integer) = nothing
+    h(a::Integer, b::Integer) = a - b
+
+    # Factor out powers of two
+    tz_num = trailing_zeros(num)
+    tz_den = trailing_zeros(den)
+    dexp = tz_num - tz_den
+    c = rational_to_float_components_impl(
+        num >> tz_num, den >> tz_den, bit_width, T, romo, h(exp_lower_bound, dexp))
+    FloatComponentsResult(
+        c.mantissa, c.exponent + dexp, inexact = c.inexact, underflowed = c.underflowed)
```

Comment:
```suggestion
    # Factor out powers of two
    tz_num = trailing_zeros(num)
    tz_den = trailing_zeros(den)
    dexp = tz_num - tz_den
    exp_lb = isnothing(exp_lower_bound) ? nothing : exp_lower_bound - dexp
    
    c = rational_to_float_components_impl(
        num >> tz_num, den >> tz_den, bit_width, T, romo, exp_lb)
    FloatComponentsResult(
        c.mantissa, c.exponent + dexp, inexact = c.inexact, underflowed = c.underflowed)
```

The branch will likely be optimized away.